subsys/caf/modules/sensor_data_aggregator.c

```diff
@@ -20,56 +20,51 @@ LOG_MODULE_REGISTER(MODULE, CONFIG_CAF_SENSOR_DATA_AGGREGATOR_LOG_LEVEL);
-	[i].sensor_descr = DT_INST_PROP(i, sensor_descr),				\
-	[i].values_in_sample = DT_INST_PROP(i, sample_size),				\
-	[i].buf_count = DT_INST_PROP(i, buf_count),					\
-	[i].buf_len = DT_INST_PROP(i, buf_data_length),					\
-	[i].agg_buffers = (struct aggregator_buffer *) &agg_ ## i ## _bufs,		\
-	[i].active_buf = (struct aggregator_buffer *) &agg_ ## i ## _bufs,
-
-#define __DEFINE_AGG							\
-	static struct aggregator aggregators[] = {			\
-		DT_INST_FOREACH_STATUS_OKAY(__DEFINE_AGGREGATOR)	\
+/* The name of the structure array that describes the aggregator buffers */
+#define __AGG_BUFFS_NAME(agg_node) DT_CAT3(agg_, agg_node, _buffs)
+
+/* This macros are used only if no memory region is used and the aggregator buffers are created
+ * in BSS.*/
+#define __DATA_BUFF_NAME(agg_node, n) DT_CAT5(agg_, agg_node,  _buff_,  n,  _data)
+#define __DEFINE_DATA(n, agg_node, size)                         \
+	static struct sensor_value __DATA_BUFF_NAME(agg_node, n) \
+		[ceiling_fraction(size, sizeof(struct sensor_value))]
+
+
+#define __INITIALIZE_BUFF(n, agg_node)                                                        \
+	COND_CODE_1(DT_NODE_HAS_PROP(agg_node, memory_region),                                \
+		({(struct sensor_value *) (DT_REG_ADDR(DT_PHANDLE(agg_node, memory_region)) + \
+			n * (DT_PROP(agg_node, buf_data_length)))}),                          \
+		({__DATA_BUFF_NAME(agg_node, n)})                                             \
+	)
+
+#define __XDEFINE_BUF_DATA(agg_node)                                                    \
+	COND_CODE_0(DT_NODE_HAS_PROP(agg_node, memory_region),                          \
+		(LISTIFY(DT_PROP(agg_node, buf_count), __DEFINE_DATA, (;),              \
+			agg_node, DT_PROP(agg_node, buf_data_length));),                \
+		()                                                                      \
+	)                                                                               \
+	static struct aggregator_buffer __AGG_BUFFS_NAME(agg_node)[] = {                \
+		LISTIFY(DT_PROP(agg_node, buf_count), __INITIALIZE_BUFF, (,), agg_node) \
+	};                                                                              \
+	BUILD_ASSERT((DT_PROP(agg_node, buf_data_length) %                              \
+		(DT_PROP(agg_node, sample_size) * sizeof(struct sensor_value))) == 0,   \
+		"Wrong sensor data or buffer size in " DT_NODE_FULL_NAME(agg_node));
+
+#define __DEFINE_BUF_DATA(i) __XDEFINE_BUF_DATA(DT_DRV_INST(i))
+
+
+#define __DEFINE_AGGREGATOR(i)                               \
+	[i].sensor_descr = DT_INST_PROP(i, sensor_descr),    \
+	[i].values_in_sample = DT_INST_PROP(i, sample_size), \
+	[i].buf_count = DT_INST_PROP(i, buf_count),          \
+	[i].buf_len = DT_INST_PROP(i, buf_data_length),      \
+	[i].agg_buffers = __AGG_BUFFS_NAME(DT_DRV_INST(i)),  \
+	[i].active_buf  = __AGG_BUFFS_NAME(DT_DRV_INST(i)),
+
+#define __DEFINE_AGG                                             \
```

Comment:
We have duplicate of code (leftover from previous commits )See line 87. Lets remove one of those.

Reply:
Ok. The __DEFINE_AGG macro removed.